Fixed memory leak and some warnings

oyvindln opened the pull request:
Fixed a memory leak.
Started adding more compiler warnings ("-Wall" leaves out many, see https://stackoverflow.com/questions/5088460/flags-to-enable-thorough-and-verbose-g-warnings/9862800#9862800 ) and added SYSTEM to include directories to avoid having warnings trigger in header files. This also allow us to use "-Werror". 
New warning are disabled for now, as many of them will trigger, my though is that we can incrementally add them and fix the occuring warnins.

Commit:
Fixed memory leak and some warnings

hwoarangmy commented:
I've tested and it seems Random.cpp is missing a #include <ctime>

Apart from that, I've seen no problem with compilation. I still have the same warnings I had before. After adding the missing include, this PR is mergeable for me. I will let @akien-mga and @Bertram25 test too because we know cmakefile is pretty sensible ^^
Moreover, I already know that the included debug CEGUI library will break @Bertram25 env. But last time, he said he would try to find why ^^

oyvindln commented:
Ahh, removing ctime was unintentional. I will make a new commit in a bit.

Commit:
readded include, cegui cmake script and linking

Added back include which was accidentally removed in last commit
Updated FindCEGUI.cmake and updated CMakeLists.txt accordingly
 - should now work fine with debug
Added conditional so plugins_d is only used on windows
Started change to client to use vector of objects instead of pointers

akien-mga commented:
It builds fine here (release build), I'll be testing soon.

akien-mga commented:
It works fine too :-)

Bertram25 commented on CMakeLists.txt:
```diff
@@ -340,16 +349,16 @@ find_package(OGRE REQUIRED)
 
-if ((CEGUI_VERSION_MINOR LESS 8) OR (CEGUI_VERSION_MAJOR LESS 1))
-    message(SEND_ERROR "CEGUI version >= 0.8.0 required")
+if("${CEGUI_VERSION}" VERSION_LESS "0.8.0")
```
Ah? I didn't know about VERSION_\* . good to know. :)

Bertram25 commented on CMakeLists.txt:
```diff
@@ -397,25 +406,22 @@ include_directories(
     #Angelscript includes
-    ${ANGELSCRIPT_DIR}/angelscript/include
-    ${ANGELSCRIPT_SRC_DIR}
+    SYSTEM ${ANGELSCRIPT_DIR}/angelscript/include
```
What is `SYSTEM`?

oyvindln replied:
> added SYSTEM to include directories to avoid having warnings trigger in header files

Bertram25 replied:
Ah ok, just found it:
http://www.cmake.org/cmake/help/v3.0/command/include_directories.html

Bertram25 commented on CMakeLists.txt:
```diff
@@ -439,6 +445,11 @@ endif()
 # Create the angelscript library file
 add_library(${AS_LIBRARY_NAME} ${AS_SOURCEFILES})
 
+# Treat warnings as errors for main build, but not angelscript, since it would fail.
```
Cool, it will help.

Bertram25 commented on source/network/ODClient.cpp:
```diff
@@ -879,6 +879,7 @@ void ODClient::processClientNotifications()
                 sendToServer(event->mPacket);
                 break;
         }
+        delete event;
```
Nice catch. :)

Bertram25 commented on source/utils/Random.cpp:
```diff
@@ -36,58 +30,35 @@
 {
     myRandomSeed = myRandomSeed * 1103515245 + 12345;
-    unsigned long returnVal = (unsigned int) (myRandomSeed / 65536) % MAX;
+    //TODO: What is the purpose of the cast?
```
Avoid returning decimal values?

oyvindln replied:
It already does integer division without the cast.

Bertram25 replied:
Then, IIRC, when A / B, B is promoted to the type of A. This means 65536 is promoted as an unsigned long and maybe the cast does put this back to an unsigned int afterwards to play fine with `% MAX.` or have better computation performance maybe?
Otherwise, I have no idea why.

hwoarangmy replied:
I can see 2 possible reasons:
- To strip the high bits. Maybe is it better for the algorithm to have only 32 bits (on most architectures)
- returnVal used to be an unsigned int and there was a cast to avoid a warning. Then, when returnVal  became unsigned long, because no warning was triggered from smaller to bigger type, it stayed unseen.

Bertram25 replied:
> returnVal used to be an unsigned int and there was a cast to avoid a warning. Then, when returnVal became unsigned long, because no warning was triggered from smaller to bigger type, it stayed unseen.

It must be that then. Casts are rarely so clever without a comment, or at least I hope. ;)
I'd propose to remove that cast and see whether it shows any problem while we're in early 0.5.0 development. WDYT guys?

Bertram25 commented on source/utils/ResourceManager.cpp:
```diff
@@ -49,7 +49,7 @@
 
 template<> ResourceManager* Ogre::Singleton<ResourceManager>::msSingleton = 0;
-#if defined(OD_DEBUG)
+#if OGRE_PLATFORM == OGRE_PLATFORM_WIN32 && defined(OD_DEBUG)
```
Here, why couldn't the application be compiled in debug mode on linux?
I do think this should be the same behaviour for every platform.
Thus, I'd remove the `OGRE_PLATFORM == OGRE_PLATFORM_WIN32 &&` part again and would remove the confusing comment as well.

What do you think?

Bertram25 replied:
> Because the plugins doesn't seem have the _d suffix on linux systems. (As the debug info is separate). They do in the windows SDK.

I just saw that. Nevermind. Funnily, other's comments have moved in the PR only, so I missed them.

Bertram25 commented:
Looks fine. :)

I'll test asap.

> Moreover, I already know that the included debug CEGUI library will break @Bertram25 env. But last time, he said he would try to find why ^^

Yeah, if the problem happens again, I'll deal with it on my own. :)

Bertram25 commented:
Ah, also, I do think this PR should be merged before mine once tested. (And I'll rebase if needed.)

akien-mga commented:
It seems everyone is happy with this PR :-)
